# Exploring Ebay Car Sales Data
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The original dataset can be found [here](https://data.world/data-society/used-cars-data). This version of the dataset is a sample of less-cleaned version of 50,000 data points that was prepared by Dataquest.

The aim of this project is to clean the data and analyze the included used car listings.

In [203]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [204]:
autos.info()
autos.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Given dataset contains 20 columns, some columns have null values. There are some columns that contain dates stored as strings. 

Also, the column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's start by cleaning the column names to make the data easier to work with.

## Cleaning Column Names

In [205]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [206]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pic_number', 'postal_code',
       'last_seen']

In [207]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  pic_number  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Initial Exploration and Cleaning
Now let's do some basic data exploration to determine what other cleaning tasks need to be done.

In [208]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pic_number   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Below are some obvious observations:
* "pic_number", "seller", "offer_type" columns can be dropped, cause all (or nearly all) of the values are the same.
* "registration_year" has strange min and max values, which are probably mistakes. 
* "price" and "odometer" columns are defined as a text. 

In [209]:
autos = autos.drop(columns = ["pic_number", "seller", "offer_type"])

In [210]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype("float") 
autos.rename({"price":"price_$"}, axis=1, inplace=True) 

In [211]:
autos.describe()

price_$  registration_year      power_ps  registration_month  \
count  5.000000e+04       50000.000000  50000.000000        50000.000000   
mean   9.840044e+03        2005.073280    116.355920            5.723360   
std    4.811044e+05         105.712813    209.216627            3.711984   
min    0.000000e+00        1000.000000      0.000000            0.000000   
25%    1.100000e+03        1999.000000     70.000000            3.000000   
50%    2.950000e+03        2003.000000    105.000000            6.000000   
75%    7.200000e+03        2008.000000    150.000000            9.000000   
max    1.000000e+08        9999.000000  17700.000000           12.000000   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [212]:
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype("float")
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True) 

In [213]:
autos.describe()

price_$  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        50000.000000  50000.000000  
mean             5.723360  50813.627300  
std              3.711984  25779.747957  
min              0.000000   1067.000000  
25%              3.000000  30451.000000  
50%              6.000000  49577.000000  
75%              9.000000  71540.000000  
max             12.000000  99998.000000

## Exploring the Odometer and Price Columns
Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns.

In [214]:
autos["odometer_km"].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

As can be seen from the output, the values in this column are rounded, which might show that sellers had to choose from pre-set options. Moreover, there are more high mileage than low mileage vehicles.

In [215]:
print(autos["price_$"].unique().shape)
print(autos["price_$"].describe())
autos["price_$"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64


0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
1000.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
999.0      434
750.0      433
900.0      420
650.0      419
850.0      410
700.0      395
4500.0     394
300.0      384
2200.0     382
950.0      379
Name: price_$, dtype: int64

The prices in this column also seem rounded. There are 2357 unique values, looks like people tend to round prices on the site.

There are 1,421 cars listed with $0 price - we might consider removing these rows (only 2% of the total number). The maximum price is one hundred million dollars, which is a lot, let's look at the highest prices further.

In [216]:
autos["price_$"].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_$, dtype: int64

In [217]:
autos["price_$"].value_counts().sort_index(ascending=True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price_$, dtype: int64

There are a small number of rows with very high values (14 at around $1 million).

eBay is an auction site, there could be items where the opening price is $1. We will keep the $1 items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to unrealistic numbers.

In [218]:
autos = autos[autos["price_$"].between(1,351000)]
autos["price_$"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

## Exploring the date columns
There are 5 columns that should represent date values:
* date_crawled
* registration_month
* registration_year
* ad_created
* last_seen

Now, the date_crawled, last_seen, and ad_created columns are all identified as string values. We need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

In [219]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [220]:
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The date_crawled dates are basically from March till April. The distribution of listings crawled on each day is roughly uniform.

In [221]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

The ad_created dates are basically from 2015-06-11 till 2016-04-07 (large variety in comparison with date_crawled).
Most dates are in 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [222]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The last three days (05/06/07-04-2016) contain a disproportionate amount of 'last seen' values. Probably these values shows the crawling period ending and don't indicate car sales.

In [223]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Registration year indicates the age of the car. However, there are some odd values in the column. The minimum value is 1000 (long before cars were invented), and the maximum is 9999 (future).

## Dealing with Incorrect Registration Year Data
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [224]:
(~autos["registration_year"].between(1900,2016)).sum()/len(autos)

0.038793369710697

As we can see, the value is less than 4%, we will remove these rows from the dataset.

In [225]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

Most of the cars were registered in the last 20 years.

## Exploring Price by Brand

In [226]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [227]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

The most popular ones are Volkswagen, BMW and Opel. Top 20 brands will be analyzed further. 

In [228]:
brand_price = {}
brands = autos["brand"].value_counts().index[:20]
for i in brands:
    mean_price = round(autos[autos["brand"]==i]["price_$"].mean())
    brand_price[i] = mean_price
brand_price

{'audi': 9337.0,
 'bmw': 8333.0,
 'citroen': 3779.0,
 'fiat': 2814.0,
 'ford': 3749.0,
 'hyundai': 5365.0,
 'mazda': 4113.0,
 'mercedes_benz': 8628.0,
 'mini': 10613.0,
 'nissan': 4743.0,
 'opel': 2975.0,
 'peugeot': 3094.0,
 'renault': 2475.0,
 'seat': 4397.0,
 'skoda': 6368.0,
 'smart': 3580.0,
 'sonstige_autos': 12339.0,
 'toyota': 5167.0,
 'volkswagen': 5402.0,
 'volvo': 4947.0}

Our observation:

* Audi, BMW and Mercedes Benz are the most expensive ones;
* Fiat and Opel are less expensive;
* Volkswagen is in the middle - this may explain its popularity.

## Exploring Mileage

In [229]:
bmp_series = pd.Series(brand_price)
mean_price = pd.DataFrame(bmp_series, columns=["mean_price"])
mean_price

mean_price
audi                9337.0
bmw                 8333.0
citroen             3779.0
fiat                2814.0
ford                3749.0
hyundai             5365.0
mazda               4113.0
mercedes_benz       8628.0
mini               10613.0
nissan              4743.0
opel                2975.0
peugeot             3094.0
renault             2475.0
seat                4397.0
skoda               6368.0
smart               3580.0
sonstige_autos     12339.0
toyota              5167.0
volkswagen          5402.0
volvo               4947.0

In [230]:
brand_mean_mileage = {}

for brand in brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_price).sort_values(ascending=False)

In [231]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
volvo                 138067
bmw                   132572
mercedes_benz         130788
opel                  129310
audi                  129157
volkswagen            128707
renault               128071
peugeot               127153
mazda                 124464
ford                  124266
seat                  121131
citroen               119694
nissan                118330
fiat                  117121
toyota                115944
skoda                 110848
hyundai               106442
smart                  99326
sonstige_autos         89956
mini                   88105

In [232]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
volvo                 138067      4947.0
bmw                   132572      8333.0
mercedes_benz         130788      8628.0
opel                  129310      2975.0
audi                  129157      9337.0
volkswagen            128707      5402.0
renault               128071      2475.0
peugeot               127153      3094.0
mazda                 124464      4113.0
ford                  124266      3749.0
seat                  121131      4397.0
citroen               119694      3779.0
nissan                118330      4743.0
fiat                  117121      2814.0
toyota                115944      5167.0
skoda                 110848      6368.0
hyundai               106442      5365.0
smart                  99326      3580.0
sonstige_autos         89956     12339.0
mini                   88105     10613.0

As clearly seen from the above table, the range of car mileages does not vary as much as the prices.

## Conclusion
In this project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings.